server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java

```diff
@@ -171,34 +172,36 @@ public Query rewrite(Query original) throws IOException {
 
     private void executeKnnVectorQuery(SearchContext context) throws IOException {
         SearchSourceBuilder source = context.request().source();
-        if (source == null || source.knnSearch() == null) {
+        if (source == null || source.knnSearch().isEmpty()) {
             return;
         }
 
         SearchExecutionContext searchExecutionContext = context.getSearchExecutionContext();
-        KnnSearchBuilder knnSearch = context.request().source().knnSearch();
-        KnnVectorQueryBuilder knnVectorQueryBuilder = knnSearch.toQueryBuilder();
+        List<KnnSearchBuilder> knnSearch = context.request().source().knnSearch();
+        List<KnnVectorQueryBuilder> knnVectorQueryBuilders = knnSearch.stream().map(KnnSearchBuilder::toQueryBuilder).toList();
 
         if (context.request().getAliasFilter().getQueryBuilder() != null) {
-            knnVectorQueryBuilder.addFilterQuery(context.request().getAliasFilter().getQueryBuilder());
+            for (KnnVectorQueryBuilder knnVectorQueryBuilder : knnVectorQueryBuilders) {
+                knnVectorQueryBuilder.addFilterQuery(context.request().getAliasFilter().getQueryBuilder());
+            }
         }
-
-        Query query = searchExecutionContext.toQuery(knnVectorQueryBuilder).query();
-        TopScoreDocCollector topScoreDocCollector = TopScoreDocCollector.create(knnSearch.k(), Integer.MAX_VALUE);
-        Collector collector = topScoreDocCollector;
-
-        if (context.getProfilers() != null) {
-            InternalProfileCollector ipc = new InternalProfileCollector(
-                topScoreDocCollector,
-                CollectorResult.REASON_SEARCH_TOP_HITS,
-                List.of()
-            );
-            context.getProfilers().getDfsProfiler().setCollector(ipc);
-            collector = ipc;
+        List<DfsKnnResults> knnResults = new ArrayList<>(knnVectorQueryBuilders.size());
+        for (int i = 0; i < knnSearch.size(); i++) {
+            Query knnQuery = searchExecutionContext.toQuery(knnVectorQueryBuilders.get(i)).query();
+            TopScoreDocCollector topScoreDocCollector = TopScoreDocCollector.create(knnSearch.get(i).k(), Integer.MAX_VALUE);
+            Collector collector = topScoreDocCollector;
+            if (context.getProfilers() != null) {
```

Comment:
What does profiling look like now?

Reply:
Oh snap, even main is giving a failure here when profiling is true:
```
» [2022-12-16T08:51:32,464][WARN ][o.e.d.FileBasedSeedHostsProvider] [runTask-0] expected, but did not find, a dynamic hosts list at [/Users/benjamintrent/Projects/elasticsearch/build/testclusters/runTask-0/config/unicast_hosts.txt]
» [2022-12-16T08:52:26,559][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [runTask-0] fatal error in thread [elasticsearch[runTask-0][search][T#1]], exiting java.lang.AssertionError: #start call misses a matching #stop call
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.profile.Timer.start(Timer.java:37)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.profile.dfs.DfsProfiler.startTimer(DfsProfiler.java:48)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.dfs.DfsPhase$1.createWeight(DfsPhase.java:115)
»       at org.apache.lucene.core@9.4.2/org.apache.lucene.search.BooleanWeight.<init>(BooleanWeight.java:59)
»       at org.apache.lucene.core@9.4.2/org.apache.lucene.search.BooleanQuery.createWeight(BooleanQuery.java:237)
»       at org.apache.lucene.core@9.4.2/org.apache.lucene.search.BoostQuery.createWeight(BoostQuery.java:124)
»       at org.apache.lucene.core@9.4.2/org.apache.lucene.search.IndexSearcher.createWeight(IndexSearcher.java:872)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.dfs.DfsPhase$1.createWeight(DfsPhase.java:119)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.dfs.DfsPhase.collectStatistics(DfsPhase.java:148)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.dfs.DfsPhase.execute(DfsPhase.java:48)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeDfsPhase(SearchService.java:452)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.lambda$executeDfsPhase$0(SearchService.java:441)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:50)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:47)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:72)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:917)
»       at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
»       at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
»       at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
»       at java.base/java.lang.Thread.run(Thread.java:1589)
»
```

Looks like we are handling profiler incorrect here.